# 01 Data Check

This notebook loads the 2025 Stack Overflow Developer Survey and performs an initial check of the columns needed for the capstone project.

The goal is to confirm:
- dataset shape
- key columns
- missing values
- target variables
- learning pathway variables
- whether the data can support the planned hypothesis test and models

In [1]:
#Tools
import pandas as pd
import numpy as np

In [2]:
#Load the dataset

df_2025 = pd.read_csv("../data/raw/results2025.csv")

df_2025.shape

C:\Users\patri\AppData\Local\Temp\ipykernel_66620\344673194.py:3: DtypeWarning: Columns (56,74,92,97,98,105,109,110,132,162,165) have mixed types. Specify dtype option on import or set low_memory=False.
  df_2025 = pd.read_csv("../data/raw/results2025.csv")


(49191, 172)

The 2025 Stack Overflow Developer Survey contains 49,191 responses and 172 columns. This notebook checks whether the dataset contains the variables needed to study learning pathways, job satisfaction, and career change consideration.

In [3]:
df_2025.head()

,ResponseId,MainBranch,Age,EdLevel,Employment,EmploymentAddl,WorkExp,LearnCodeChoose,LearnCode,LearnCodeAI,...,AIAgentOrchestration,AIAgentOrchWrite,AIAgentObserveSecure,AIAgentObsWrite,AIAgentExternal,AIAgentExtWrite,AIHuman,AIOpen,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,25-34 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Employed,"Caring for dependents (children, elderly, etc.)",8.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,Vertex AI,NaN,NaN,NaN,ChatGPT,NaN,When I don’t trust AI’s answers,"Troubleshooting, profiling, debugging",61256.0,10.0
1,2,I am a developer by profession,25-34 years old,"Associate degree (A.A., A.S., etc.)",Employed,NaN,2.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers;When I want to...,All skills. AI is a flop.,104413.0,9.0
2,3,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Independent contractor, freelancer, or self-em...",None of the above,10.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code;GitHub Copilot;Google Gemini,NaN,When I don’t trust AI’s answers;When I want to...,"Understand how things actually work, problem s...",53061.0,8.0
3,4,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Employed,None of the above,4.0,"Yes, I am not new to coding but am learning ne...","Other online resources (e.g. standard search, ...","Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code,NaN,When I don’t trust AI’s answers;When I want to...,NaN,36197.0,6.0
4,5,I am a developer by profession,35-44 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)","Independent contractor, freelancer, or self-em...","Caring for dependents (children, elderly, etc.)",21.0,"No, I am not new to coding and did not learn n...",NaN,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers,"critical thinking, the skill to define the tas...",60000.0,7.0


The 2025 Stack Overflow Developer Survey contains 49,191 responses and 172 columns. The full dataset includes demographic information, education, learning pathways, employment characteristics, compensation, AI attitudes, job satisfaction, and career change consideration.

In [4]:
project_cols = [
    "ResponseId",
    "JobSat",
    "NewRole",
    "LearnCodeChoose",
    "LearnCode",
    "LearnCodeAI",
    "EdLevel",
    "YearsCode",
    "WorkExp",
    "ConvertedCompYearly",
    "Age",
    "Employment",
    "DevType",
    "OrgSize",
    "RemoteWork",
    "Country",
    "AISelect",
    "AIThreat"
]

df_projresult25 = df_2025[project_cols].copy()

df_projresult25.head()

,ResponseId,JobSat,NewRole,LearnCodeChoose,LearnCode,LearnCodeAI,EdLevel,YearsCode,WorkExp,ConvertedCompYearly,Age,Employment,DevType,OrgSize,RemoteWork,Country,AISelect,AIThreat
0,1,10.0,I have neither consider or transitioned into a...,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...","Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",14.0,8.0,61256.0,25-34 years old,Employed,"Developer, mobile",20 to 99 employees,Remote,Ukraine,"Yes, I use AI tools monthly or infrequently",I'm not sure
1,2,9.0,I have transitioned into a new career and/or i...,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...","Associate degree (A.A., A.S., etc.)",10.0,2.0,104413.0,25-34 years old,Employed,"Developer, back-end",500 to 999 employees,"Hybrid (some in-person, leans heavy to flexibi...",Netherlands,"Yes, I use AI tools weekly",I'm not sure
2,3,8.0,I have transitioned into a new career and/or i...,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...","Bachelor’s degree (B.A., B.S., B.Eng., etc.)",12.0,10.0,53061.0,35-44 years old,"Independent contractor, freelancer, or self-em...","Developer, front-end",NaN,NaN,Ukraine,"Yes, I use AI tools daily",No
3,4,6.0,I have neither consider or transitioned into a...,"Yes, I am not new to coding but am learning ne...","Other online resources (e.g. standard search, ...","Yes, I learned how to use AI-enabled tools for...","Bachelor’s degree (B.A., B.S., B.Eng., etc.)",5.0,4.0,36197.0,35-44 years old,Employed,"Developer, back-end","10,000 or more employees",Remote,Ukraine,"Yes, I use AI tools weekly",No
4,5,7.0,I have neither consider or transitioned into a...,"No, I am not new to coding and did not learn n...",NaN,"Yes, I learned how to use AI-enabled tools for...","Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",22.0,21.0,60000.0,35-44 years old,"Independent contractor, freelancer, or self-em...",Engineering manager,NaN,NaN,Ukraine,"Yes, I use AI tools weekly",No


In [5]:
df_projresult25.shape

(49191, 18)

The missing value counts below indicate that missing values are significant and a strategy needs to be developed to handle them

In [6]:
missing = (
    df_projresult25.isna()
    .sum()
    .sort_values(ascending=False)
    .to_frame("missing_count")
)

missing["missing_percent"] = (
    missing["missing_count"] / len(df_projresult25) * 100
).round(1)

missing

,missing_count,missing_percent
ConvertedCompYearly,25244,51.3
JobSat,22521,45.8
LearnCode,15635,31.8
AISelect,15471,31.5
RemoteWork,15411,31.3
OrgSize,15013,30.5
Country,13754,28.0
NewRole,13664,27.8
AIThreat,13113,26.7
WorkExp,6298,12.8


the next cell creates the target variable of high job satisfaction where job sat of greater than 8 is 1. Anything lower is 0.

In [7]:
df_projresult25["high_job_sat"] = np.where(
    df_projresult25["JobSat"] >= 8,
    1,
    0
)

df_projresult25.loc[df_projresult25["JobSat"].isna(), "high_job_sat"] = np.nan

The next cell creates the target variable for considering a career change

In [8]:
df_projresult25["career_change_considered"] = np.where(
    df_projresult25["NewRole"].str.contains("neither", case=False, na=False),
    0,
    1
)

df_projresult25.loc[df_projresult25["NewRole"].isna(), "career_change_considered"] = np.nan

df_projresult25["career_change_considered"].value_counts(dropna=False)

career_change_considered
1.0    19322
0.0    16205
NaN    13664
Name: count, dtype: int64

If the answer contains the word "neither", code it as 0 because they did not consider or transition. Otherwise, code it as 1 because they either considered a change or already transitioned. Then put missing answers back to missing.

If somebody did not answer the job satisfaction question, they cannot be considered for the job satisfaction model.

In [9]:
df_jobsat_model = df_projresult25.dropna(subset=["JobSat"]).copy()

If somebody did not answer the career change question, they cannot be considered for the classification model

In [10]:
df_newrole_model = df_projresult25.dropna(subset=["NewRole"]).copy()

In [11]:
categorical_cols = [
    "LearnCodeChoose",
    "LearnCode",
    "LearnCodeAI",
    "EdLevel",
    "Age",
    "Employment",
    "DevType",
    "OrgSize",
    "RemoteWork",
    "Country",
    "AISelect",
    "AIThreat"
]

In [12]:
df_projresult25["active_learner"] = np.where(
    df_projresult25["LearnCodeChoose"].str.contains(
        "not new to coding but am learning",
        case=False,
        na=False
    ),
    1,
    0
)

df_projresult25.loc[
    df_projresult25["LearnCodeChoose"].isna()
    | df_projresult25["LearnCodeChoose"].eq("Missing"),
    "active_learner"
] = np.nan

df_projresult25["active_learner"].value_counts(dropna=False)

active_learner
1.0    32362
0.0    14496
NaN     2333
Name: count, dtype: int64

Missingness may actually be meaningful.
It avoids dropping thousands of rows.
One-hot encoding can handle "Missing" later.

In [13]:
for col in categorical_cols:
    df_projresult25[col] = df_projresult25[col].fillna("Missing")

Because salary and experience can have extreme outliers. Median is safer.

In [14]:
df_projresult25["log_salary"] = np.log1p(df_projresult25["ConvertedCompYearly"])

the cell above with log_salary is a transformed version of salary. It shrinks very large salary values so extreme outliers do not overpower the model. 

Compensation was transformed using a natural log transformation to reduce the influence of extreme salary outliers while preserving the relative ordering of salaries.

In [15]:
df_projresult25["WorkExp_clean"] = pd.to_numeric(
    df_projresult25["WorkExp"],
    errors="coerce"
)

In [16]:
numeric_cols = [
    "YearsCode_clean",
    "WorkExp_clean",
    "log_salary"
]

In [17]:
def clean_years_code(value):
    if pd.isna(value):
        return np.nan
    if value == "Less than 1 year":
        return 0.5
    if value == "More than 50 years":
        return 51
    return float(value)

df_projresult25["YearsCode_clean"] = df_projresult25["YearsCode"].apply(clean_years_code)

In [18]:
for col in numeric_cols:
    df_projresult25[col] = df_projresult25[col].fillna(df_projresult25[col].median())

In [19]:
df_projresult25[["YearsCode", "YearsCode_clean", "WorkExp", "WorkExp_clean", "ConvertedCompYearly", "log_salary"]].head()

,YearsCode,YearsCode_clean,WorkExp,WorkExp_clean,ConvertedCompYearly,log_salary
0,14.0,14.0,8.0,8.0,61256.0,11.022833
1,10.0,10.0,2.0,2.0,104413.0,11.556119
2,12.0,12.0,10.0,10.0,53061.0,10.879216
3,5.0,5.0,4.0,4.0,36197.0,10.496759
4,22.0,22.0,21.0,21.0,60000.0,11.002117


We need to save this data to a csv for other analysis


In [20]:
df_projresult25.to_csv(
    "../data/processed/stackoverflow_2025_project_cleaned.csv",
    index=False
)

In [21]:
df_projresult25.columns.tolist()

['ResponseId',
 'JobSat',
 'NewRole',
 'LearnCodeChoose',
 'LearnCode',
 'LearnCodeAI',
 'EdLevel',
 'YearsCode',
 'WorkExp',
 'ConvertedCompYearly',
 'Age',
 'Employment',
 'DevType',
 'OrgSize',
 'RemoteWork',
 'Country',
 'AISelect',
 'AIThreat',
 'high_job_sat',
 'career_change_considered',
 'active_learner',
 'log_salary',
 'WorkExp_clean',
 'YearsCode_clean']

The original compensation column was preserved for exploratory analysis. A log-transformed salary variable was created for modeling to reduce the influence of extreme salary outliers.

In [22]:
df_projresult25[["ConvertedCompYearly", "log_salary"]].head()

,ConvertedCompYearly,log_salary
0,61256.0,11.022833
1,104413.0,11.556119
2,53061.0,10.879216
3,36197.0,10.496759
4,60000.0,11.002117


## Data Check Summary

The 2025 Stack Overflow survey supports this project. The cleaned dataset includes engineered variables for job satisfaction, career change consideration, active learning status, learning pathway flags, cleaned experience columns, and log-transformed salary.

Two important data-quality decisions were made:
- `NewRole` was converted into a binary career change consideration variable.
- `LearnCodeChoose` was converted into an active learner variable.

The cleaned file was saved to `data/processed/stackoverflow_2025_project_cleaned.csv` for use in EDA and modeling.